MIP21: fix remediation period bug

e18r opened the pull request:
This PR fixes a bug in the remediation period checks. Also it includes some test improvements, two new tests, the renaming of `fbo` to `outputConduit` in `RwaUrn.sol`, the addition of some shortened revert strings in the oracle, and verbose test matching.

e18r committed:
rename fbo to outputConduit

e18r committed:
test improvements

* added a few checks
* renamed some tests
* added two more tests

hermetic-himbo committed:
fix comparison direction bug in good() and cull(), fix tests

livnev commented:
Hey @e18r I love the improved and more detailed tests!

hermetic-himbo committed:
RwaLiquidationOracle.sol: add revert strings

livnev commented:
Hey @e18r, I hear from @godsflaw that a bug was already suspected here, which I realised also from looking at the tests you added. The direction of the timestamp comparisons in `cull` and `good` was the wrong way around. To clarify, the way the remediation period works is:
(1) If something is wrong, governance sets DC to 0 and calls `tell`, which kicks off the timer on the remeditation period (in the tests this is set to 2 weeks). In the mean time, the borrower cannot more because the DC is set to 0, but calling `good` return `true` because the loan is not yet delinquent while it is still in the remediation phase.
(2) The borrower has until the end of the remediation period to make changes or provide information satisfactory to governance that the loan is OK. If they do so and governance is satisfied, then governance calls `cure` and resets the DC back to where it should be.
(3) If they don't or governance is not satisfied, then after the remediation period has elapsed, governance may call `cull` to actually liquidate the loan. At this point `good` also begins to return `false`, indicating to off-chain actors such as trustees that the loan is delinquent and that they should proceed with off-chain liquidation of the loan.

It looks like maybe the error created some confusion about the intention here so I thought to write that out in case it wasn't obvious. I've now corrected the code and also corrected the tests to properly test for this (earlier some of them were actually relying on the wrong behaviour).     

EDIT: an earlier version of this message had said "right way around" instead of "wrong way around". That was also a confusing error that I've since corrected, sorry.

livnev commented:
@godsflaw @gbalabasquer @wilbarnes @brianmcmichael Could one of you guys review this PR and then I will merge it

e18r committed:
time comparison improvements

* when comparing the current time with a date, put the current time first
* in tests, warp to 3 weeks instead of 2 to ensure the remediation
period has elapsed

e18r committed:
prevent revert strings from exceeding 32 chars

e18r committed:
add verbose test matching option to `make test`

e18r commented:
Hi @livnev, thank you for the explanation. We hadn't realized this behavior was a bug, which led us into thinking that the whole remediation period stuff was unnecessary. Hence PR #13, which I'm now going to close in favor of this one.

I made some minor changes that I think can help prevent such bugs in the future:

* I put `block.timestamp` first when comparing it to something else.
* In the tests, I changed the `hevm.warp` to 3 weeks to make sure we're past the remediation period.

Also, I added verbose test matching in `make test` and shortened the revert strings to save some deployment gas.

e18r committed:
remove unnecessary newlines from test

godsflaw commented on src/RwaLiquidationOracle.sol:
```diff
@@ -111,14 +111,14 @@ contract RwaLiquidationOracle {
-        require(ilks[ilk].pip != address(0));
-        require(add(ilks[ilk].toc, ilks[ilk].tau) >= block.timestamp);
+        require(ilks[ilk].pip != address(0), "RwaOracle/unknown-ilk");
+        require(block.timestamp >= add(ilks[ilk].toc, ilks[ilk].tau), "RwaOracle/early-cull");
```
@livnev this is what we want right?

livnev replied:
The new version is correct, yes. `cull` should only be callable once the remediation period has elapsed. Hopefully the revert message also documents the intent here a bit.

godsflaw commented on src/RwaLiquidationOracle.sol:
```diff
@@ -132,7 +132,8 @@ contract RwaLiquidationOracle {
-        return (ilks[ilk].toc == 0 || add(ilks[ilk].toc, ilks[ilk].tau) < block.timestamp);
+        require(ilks[ilk].pip != address(0), "RwaOracle/unknown-ilk");
+        // tell not called or still in remediation period
+        return (ilks[ilk].toc == 0 || block.timestamp < add(ilks[ilk].toc, ilks[ilk].tau));
```
@livnev and this too.  This is what we want?

livnev replied:
The new version is correct, yes: the position is `good()` if there is not an outstanding `tell` or if the remediation period has not elapsed yet.

livnev commented:
@e18r Great, agreed on all points. Will wait for one more review before merging this PR.

godsflaw reviewed (approved):
LGTM if @livnev is happy with the changes in the inequalities.

livnev commented:
This has been merged.
